Type checking support for host bindings

crisbeto opened the pull request:
Historically Angular's type checking only extended to templates, however host bindings can contain expressions as well which can have type checking issues of their own ([example](https://github.com/angular/components/pull/30570) and [example](https://github.com/angular/components/pull/30606)). These changes add the ability to type check the `host` object literal, `@HostBinding` and `@HostListener` decorators of `@Directive` and `@Component` classes. The new functionality is hidden behind the `typeCheckHostBindings` compiler flag, but we plan to enable it by default in a future version.

This also allows us to enable language service features in host bindings like renaming and hover information. Here's demo showing how it's integrated:

![host-bindings-demo](https://github.com/user-attachments/assets/0492c2dc-b6bd-4f7a-a98a-cc7e54303ef1)


**Note:** the language service integration for this is done, but I'll send it out in a separate PR to make this one easier to review.

crisbeto committed:
refactor(compiler): add AST node for host element

Adds the `HostElement` AST node that can be used to represent the host of a directive. Also exposes the `BindingParser` type so it's easier to pass around the return value of `makeBindingParser`.

crisbeto committed:
refactor(compiler): account for host element in binder

Updates the `R3TargetBinder` to ingest host element nodes.

crisbeto committed:
refactor(compiler-cli): track host binding information

Sets up the infrastructure to track the host bindings of directives for type checking purposes.

crisbeto committed:
refactor(compiler-cli): add logic for synthesizing host element

Adds the logic that creates a `HostElement` AST node based on the host bindings of a directive.

crisbeto committed:
refactor(compiler-cli): produce host binding type checking information for components

Sets up the logic that produces the information necessary to type check host bindings of a component. Also introduces a compiler flag for toggling checking of host bindings.

crisbeto committed:
refactor(compiler-cli): enable type checking of directive classes

Updates the directive handler to enable type checking of directives. Currently this only applies to checking of host bindings.

crisbeto commented:
Yeah I have it on my TODO to look into syntax highlighting. It's in the vscode extension repo though.

JoostK commented on packages/compiler-cli/test/ngtsc/host_bindings_type_check_spec.ts:
```diff
@@ -0,0 +1,699 @@
+  return diag.file!.text.slice(diag.start!, diag.start! + diag.length!);
+}
+
+runInEachFileSystem(() => {
```
I havent't looked at all tests in detail, but I realised there weren't any cases for `window:` and `document:` prefixes and I haven't seen those being handled specially so they're probably incorrectly targeting the host element instead of window/document.

crisbeto replied:
There's a `document:click` in the `should check host event listeners with a target` test and it seems to get inferred as a `MouseEvent` correctly.

JoostK replied:
I suppose that is because the host element for `button[foo]` also supports the `click` event, You'd likely see issues for host bindings for properties that are only available on `document`/`window`.

crisbeto replied:
How can you access `document` or `window` in the handler expression though? The only way I can think of is `$event.target`, but even then TS infers it to be `EventTarget` rather than the node type.

crisbeto replied:
Okay I see what you mean. In that case the event gets inferred as `Event`, rather than `BeforeUnloadEvent`. That being said, we don't see to handle this in the template TCB either. E.g. here's the same setup in a template:

<img width="1184" alt="image" src="https://github.com/user-attachments/assets/0ec36a13-957f-4994-a0f5-cf989b936825" />

JoostK replied:
I'd expect that the target prefix is designed to be used in host listeners, but it could be that they also (unintentionally, maybe?) work in templates?

crisbeto replied:
It's a bit counter-intuitive, but they work in templates too. E.g. you can do `<button (window:scroll)="...">` to listen to the global `scroll` event "through" some random button.

JoostK reviewed (commented):
Excited for this feature! 💯 

---

The commit message of the `feat` commit mentions:

> Note that initially the new functionality is hidden behind the `typeCheckHostBindings` compiler flag

Since the option is public, I'd rephrase that to

> Note that initially the new functionality is disabled by default and has to be enabled using the `typeCheckHostBindings` compiler flag.

devversion reviewed (approved):
Looks pretty good to me. Nice work!

I agree with a few comments from Joost, and will defer further to him, but couldn't spot anything else on my own.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/host_bindings.ts:
```diff
@@ -118,6 +124,44 @@ export function createHostElement(
+}
+
+/**
+ * Determines if a given node is a marker that descendant nodes are used to check host bindings.
```
Thinking about it more, perhaps the marker can be renamed to `hostBindingsBlockGuard` or something like that?

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/host_bindings.ts:
```diff
@@ -0,0 +1,507 @@
+  // back to the source. E.g. `@HostListener('foo') handleFoo() {}` in source code would look
+  // something like `(foo)="handleFoo()"` in the AST. Instead we construct the expressions
+  // manually here. Note that we use a dummy span with -1/-1 as offsets, because it isn't
+  // used for type checking and constructing it accurately would take some effort.
```
For context, these are the spans within the expression itself, rather than the source file. E.g. the span for `'foo'` would be `start: 0, end: 6`. It's fairly straightforward to count the number of characters for each expression, but I figured it's not worth the hassle considering that it's not used anywhere.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/host_bindings.ts:
```diff
@@ -0,0 +1,463 @@
+
+  if (args.length > 1 && ts.isArrayLiteralExpression(args[1])) {
+    for (const expr of args[1].elements) {
+      // Stop the process if we can't resolve a specific parameter.
+      if (!ts.isStringLiteralLike(expr)) {
+        return;
+      }
```
I think that's a good idea 👍 I'll turn them into `$any(null)`.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/host_bindings.ts:
```diff
@@ -0,0 +1,463 @@
+      const start = expr.getStart();
+      const end = expr.getEnd();
+      const exprSourceSpan = new AbsoluteSourceSpan(start, end);
+      argNodes.push(
+        new PropertyRead(
+          span,
+          exprSourceSpan,
+          exprSourceSpan,
+          new ImplicitReceiver(span, new AbsoluteSourceSpan(start, start)),
+          expr.text,
+        ),
+      );
```
Oh for some reason I thought that we only allowed to pass `$event` here, but you're totally right that it can be any expression. Switching it over to the parser...

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/host_bindings.ts:
```diff
@@ -0,0 +1,463 @@
+  // This will preserve the right mappings, but can lead to parsing errors, because some of the
+  // strings won't have to be escaped anymore. We could add a mode to the parser that allows it to
+  // recover from such cases, but it'll introduce more complexity that we may not want to take on.
+  // 4. Constructing some sort of string like `<host ${name.getText()}=${initializer.getText()}/>`,
+  // passing it through the HTML parser and extracting the first attribute from it - wasn't explored
+  // much, but likely has the same issues as approach #3.
```
Ah interesting. When I was looking for a way to handle this, I was comparing with templates and noticed that it was dealing with it correctly. Too bad `escapedString` is in the ML parser, rather than the expression parser. In theory we could go with approach #4 so we can take advantage of `escapedString`, but it'll also come with a bit more overhead.

crisbeto commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -777,6 +793,8 @@ class TemplateBinder extends RecursiveAstVisitor implements Visitor {
     ) {
       nodeOrNodes.children.forEach((node) => node.visit(this));
       this.nestingLevel.set(nodeOrNodes, this.level);
+    } else if (nodeOrNodes instanceof HostElement) {
+      this.nestingLevel.set(nodeOrNodes, this.level);
```
I think in practice the `level` here would've always been zero anyways. I've hardcoded it and left a comment.

crisbeto reviewed (commented):
Feedback has been addressed.

JoostK commented on packages/compiler-cli/test/ngtsc/host_bindings_type_check_spec.ts:
```diff
@@ -0,0 +1,718 @@
+
+          @Component({template: ''})
+          export class Comp {
+            @HostListener('click', [foo]) handleClick(value: string) {};
```
nit: I'd change this to e.g.

```ts
@HostListener('click', ['1', two, '3'])
handleClick(a: string, b: string, c: string) {}
```

to verify that there are diagnostics for all arguments except for the dynamic one.

crisbeto committed:
feat(compiler-cli): support type checking of host bindings

Historically Angular's type checking only extended to templates, however host bindings can contain expressions as well which can have type checking issues of their own. These changes expand the type checking infrastructure to cover the `host` object literal, `@HostBinding` decorators and `@HostListener` with full language service support coming in future commits.

Note that initially the new functionality is disabled by default and has to be enabled using the `typeCheckHostBindings` compiler flag.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

devversion reviewed (approved):
Reviewed-for: public-api

alxhub commented on packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:
```diff
@@ -430,6 +430,9 @@ export interface TargetOptions {
  */
 export interface MiscOptions {
+  /** Whether type checking of host bindings is enabled. */
+  typeCheckHostBindings?: boolean;
```
Technically this could go in `DiagnosticOptions` or `StrictTemplateOptions`. If we wanted to be really extra, we could rename the latter to `TypeCheckingOptions`. These are purely for our own documentation anyway.

crisbeto committed:
refactor(compiler-cli): move typeCheckHostBindings option

Moves the `typeCheckHostBindings` option into `StrictTemplateOptions` and renames the latter.

pkozlowski-opensource commented:
This PR was merged into the repository by commit bf2b0253bccb10c15d6394ce28e4fa6ed791b912.

The changes were merged into the following branches: main
